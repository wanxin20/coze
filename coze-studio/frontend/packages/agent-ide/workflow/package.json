{
  "name": "@coze-agent-ide/workflow",
  "version": "0.0.1",
  "description": "bot 与 workflow 交互的内容",
  "license": "Apache-2.0",
  "author": "meixuliang.3@bytedance.com",
  "maintainers": [],
  "main": "src/index.tsx",
  "scripts": {
    "build": "exit 0",
    "lint": "eslint ./ --cache",
    "test": "vitest --run --passWithNoTests",
    "test:cov": "npm run test -- --coverage"
  },
  "dependencies": {
    "@coze-agent-ide/bot-plugin-export": "workspace:*",
    "@coze-agent-ide/navigate": "workspace:*",
    "@coze-agent-ide/plugin-shared": "workspace:*",
    "@coze-agent-ide/space-bot": "workspace:*",
    "@coze-agent-ide/tool": "workspace:*",
    "@coze-agent-ide/workflow-as-agent-adapter": "workspace:*",
    "@coze-arch/bot-api": "workspace:*",
    "@coze-arch/bot-error": "workspace:*",
    "@coze-arch/bot-flags": "workspace:*",
    "@coze-arch/bot-hooks": "workspace:*",
    "@coze-arch/bot-icons": "workspace:*",
    "@coze-arch/bot-semi": "workspace:*",
    "@coze-arch/bot-studio-store": "workspace:*",
    "@coze-arch/bot-utils": "workspace:*",
    "@coze-arch/coze-design": "0.0.6-alpha.346d77",
    "@coze-arch/i18n": "workspace:*",
    "@coze-arch/idl": "workspace:*",
    "@coze-arch/report-events": "workspace:*",
    "@coze-arch/report-tti": "workspace:*",
    "@coze-common/assets": "workspace:*",
    "@coze-studio/bot-detail-store": "workspace:*",
    "@coze-studio/components": "workspace:*",
    "@coze-workflow/base": "workspace:*",
    "@coze-workflow/components": "workspace:*",
    "ahooks": "^3.7.8",
    "classnames": "^2.3.2",
    "copy-to-clipboard": "^3.3.3",
    "immer": "^10.0.3",
    "lodash-es": "^4.17.21",
    "zustand": "^4.4.7"
  },
  "devDependencies": {
    "@coze-arch/bot-env": "workspace:*",
    "@coze-arch/bot-typings": "workspace:*",
    "@coze-arch/eslint-config": "workspace:*",
    "@coze-arch/stylelint-config": "workspace:*",
    "@coze-arch/ts-config": "workspace:*",
    "@coze-arch/vitest-config": "workspace:*",
    "@testing-library/jest-dom": "^6.1.5",
    "@testing-library/react": "^14.1.2",
    "@testing-library/react-hooks": "^8.0.1",
    "@types/lodash-es": "^4.17.10",
    "@types/react": "18.2.37",
    "@types/react-dom": "18.2.15",
    "@vitest/coverage-v8": "~3.0.5",
    "react": "~18.2.0",
    "react-dom": "~18.2.0",
    "stylelint": "^15.11.0",
    "vite-plugin-svgr": "~3.3.0",
    "vitest": "~3.0.5"
  },
  "peerDependencies": {
    "react": ">=18.2.0",
    "react-dom": ">=18.2.0"
  },
  "// deps": "immer@^10.0.3 为脚本自动补齐，请勿改动"
}

