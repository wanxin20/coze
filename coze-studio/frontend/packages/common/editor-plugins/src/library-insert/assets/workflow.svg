<svg width="15" height="14" viewBox="0 0 15 14" fill="none" xmlns="http://www.w3.org/2000/svg">
<g id="Component 2">
<path id="Union" fill-rule="evenodd" clip-rule="evenodd" d="M9.06505 5.38087C9.27846 5.65615 9.61235 5.83333 9.98763 5.83333H12.6126C13.257 5.83333 13.7793 5.311 13.7793 4.66667V2.04167C13.7793 1.39733 13.257 0.875 12.6126 0.875H9.98763C9.3433 0.875 8.82096 1.39733 8.82096 2.04167V4.19269L8.80691 4.16985L6.48763 5.5971V5.25C6.48763 4.60567 5.9653 4.08333 5.32096 4.08333L2.69596 4.08333C2.05163 4.08333 1.5293 4.60567 1.5293 5.25V8.75C1.5293 9.39433 2.05163 9.91667 2.69596 9.91667H5.32096C5.9653 9.91667 6.48763 9.39433 6.48763 8.75V8.36035L8.82096 9.72859V11.9583C8.82096 12.6027 9.3433 13.125 9.98763 13.125H12.6126C13.257 13.125 13.7793 12.6027 13.7793 11.9583V9.33333C13.7793 8.689 13.257 8.16667 12.6126 8.16667H9.98763C9.64375 8.16667 9.33463 8.31544 9.12111 8.55214L6.48763 7.00789V6.96697L9.06505 5.38087ZM9.98763 2.04167V4.66667H12.6126V2.04167H9.98763ZM9.98763 9.33333V11.9583H12.6126V9.33333H9.98763ZM2.69596 8.75L2.69596 5.25L5.32096 5.25L5.32096 8.75H2.69596Z" fill="#4E40E5"/>
</g>
</svg>
