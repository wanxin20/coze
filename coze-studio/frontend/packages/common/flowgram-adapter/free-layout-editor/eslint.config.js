const { defineConfig } = require('@coze-arch/eslint-config');

module.exports = defineConfig({
  packageRoot: __dirname,
  preset: 'web',
  rules: {
    'no-cond-assign': 'off',
    '@typescript-eslint/no-non-null-assertion': 'off',
    '@typescript-eslint/no-explicit-any': 'off',
    '@coze-arch/no-batch-import-or-export': 'off',
    'max-params': 'off',
  },
});
