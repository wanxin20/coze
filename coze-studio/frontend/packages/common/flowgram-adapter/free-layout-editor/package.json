{
  "name": "@flowgram-adapter/free-layout-editor",
  "version": "0.0.1",
  "description": "对 flowgram 的封装",
  "license": "Apache-2.0",
  "author": "wencheng.lwc@bytedance.com",
  "maintainers": [],
  "exports": {
    ".": {
      "types": "./src/index.ts",
      "import": "./src/index.ts",
      "require": "./src/index.ts"
    },
    "./css-load": {
      "types": "./src/css-load.ts",
      "import": "./src/css-load.ts",
      "require": "./src/css-load.ts"
    }
  },
  "main": "src/index.ts",
  "types": "src/index.ts",
  "files": [
    "src",
    "bin"
  ],
  "scripts": {
    "build": "exit 0",
    "lint": "eslint ./ --cache",
    "test": "vitest --run --passWithNoTests",
    "test:cov": "npm run test -- --coverage"
  },
  "dependencies": {
    "@flowgram.ai/core": "0.1.28",
    "@flowgram.ai/document": "0.1.28",
    "@flowgram.ai/form": "0.1.28",
    "@flowgram.ai/form-core": "0.1.28",
    "@flowgram.ai/free-auto-layout-plugin": "0.1.28",
    "@flowgram.ai/free-container-plugin": "0.1.28",
    "@flowgram.ai/free-history-plugin": "0.1.28",
    "@flowgram.ai/free-layout-core": "0.1.28",
    "@flowgram.ai/free-layout-editor": "0.1.28",
    "@flowgram.ai/free-lines-plugin": "0.1.28",
    "@flowgram.ai/free-node-panel-plugin": "0.1.28",
    "@flowgram.ai/free-snap-plugin": "0.1.28",
    "@flowgram.ai/free-stack-plugin": "0.1.28",
    "@flowgram.ai/history-node-plugin": "0.1.28",
    "@flowgram.ai/minimap-plugin": "0.1.28",
    "@flowgram.ai/node-core-plugin": "0.1.28",
    "@flowgram.ai/node-variable-plugin": "0.1.28",
    "@flowgram.ai/renderer": "0.1.28",
    "@flowgram.ai/variable-plugin": "0.1.28",
    "reflect-metadata": "^0.1.13"
  },
  "devDependencies": {
    "@babel/core": "^7.26.0",
    "@coze-arch/eslint-config": "workspace:*",
    "@coze-arch/stylelint-config": "workspace:*",
    "@coze-arch/ts-config": "workspace:*",
    "@coze-arch/vitest-config": "workspace:*",
    "@testing-library/jest-dom": "^6.1.5",
    "@testing-library/react": "^14.1.2",
    "@testing-library/react-hooks": "^8.0.1",
    "@types/node": "^18",
    "@types/react": "18.2.37",
    "@types/react-dom": "18.2.15",
    "@vitest/coverage-v8": "~3.0.5",
    "react": "~18.2.0",
    "react-dom": "~18.2.0",
    "react-is": ">= 16.8.0",
    "styled-components": ">=4",
    "stylelint": "^15.11.0",
    "vitest": "~3.0.5"
  },
  "peerDependencies": {
    "react": ">=18.2.0",
    "react-dom": ">=18.2.0"
  }
}

