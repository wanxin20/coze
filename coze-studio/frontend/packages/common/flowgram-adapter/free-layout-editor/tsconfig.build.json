{
  "extends": "@coze-arch/ts-config/tsconfig.web.json",
  "compilerOptions": {
    "baseUrl": "./",
    "paths": {},
    "types": [],
    "jsx": "react",
    "isolatedModules": true,
    "preserveSymlinks": false,
    "strictPropertyInitialization": false,
    "module": "ESNext",
    "rootDir": "./src",
    "outDir": "./dist",
    "tsBuildInfoFile": "dist/tsconfig.build.tsbuildinfo"
  },
  "include": ["src"],
  "references": [
    {
      "path": "../../../../config/eslint-config/tsconfig.build.json"
    },
    {
      "path": "../../../../config/stylelint-config/tsconfig.build.json"
    },
    {
      "path": "../../../../config/ts-config/tsconfig.build.json"
    },
    {
      "path": "../../../../config/vitest-config/tsconfig.build.json"
    }
  ],
  "$schema": "https://json.schemastore.org/tsconfig"
}
