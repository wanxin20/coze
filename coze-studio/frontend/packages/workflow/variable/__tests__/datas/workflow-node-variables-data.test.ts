/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  WorkflowNodeInputVariablesData,
  WorkflowNodeOutputVariablesData,
  WorkflowNodeRefVariablesData,
} from 'src';
import { loopJSON } from '__tests__/workflow.mock';
import { createContainer } from '__tests__/create-container';
import { WorkflowDocument } from '@flowgram-adapter/free-layout-editor';

describe('test workflow node variables entity data', () => {
  let workflowDocument: WorkflowDocument;

  beforeEach(async () => {
    const container = createContainer();
    workflowDocument = container.get<WorkflowDocument>(WorkflowDocument);

    await workflowDocument.fromJSON(loopJSON);
  });

  test('test workflow-node-input-variables-data', () => {
    const nodes = workflowDocument.getAllNodes();

    expect(
      nodes.map(_node => {
        const data = _node.getData(WorkflowNodeInputVariablesData);

        return [
          _node.id,
          data.inputParameters,
          data.inputVariables.map(_input => _input.refVariable?.viewType),
        ];
      }),
    ).toMatchSnapshot();
  });

  test('test workflow-node-output-variables-data', () => {
    const nodes = workflowDocument.getAllNodes();

    expect(
      nodes.map(_node => {
        const data = _node.getData(WorkflowNodeOutputVariablesData);

        return [
          _node.id,
          data.variables.map(
            _variable => data.getVariableByKey(_variable.key)?.viewType,
          ),
        ];
      }),
    ).toMatchSnapshot();
  });

  test('test workflow-node-ref-variables-data', () => {
    const nodes = workflowDocument.getAllNodes();

    expect(
      nodes.map(_node => {
        const data = _node.getData(WorkflowNodeRefVariablesData);

        return [
          _node.id,
          data.refs,
          Object.values(data.refVariables).map(_ref => _ref?.viewType),
          data.hasGlobalRef,
        ];
      }),
    ).toMatchSnapshot();
  });

  test('test batchUpdateRefs', () => {
    const endNode = workflowDocument.getNode('end');

    const data = endNode?.getData(WorkflowNodeRefVariablesData);

    data?.batchUpdateRefs([
      {
        beforeKeyPath: ['llm_0'],
        afterKeyPath: ['llm_test'],
      },
      {
        beforeKeyPath: ['start', 'Object'],
        afterKeyPath: ['start', 'Drilldown', 'Object'],
      },
    ]);

    expect(data?.refs).toMatchSnapshot();
  });
});
