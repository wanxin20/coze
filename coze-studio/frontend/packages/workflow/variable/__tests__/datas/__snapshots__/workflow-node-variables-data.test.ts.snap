// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test workflow node variables entity data > test batchUpdateRefs 1`] = `
{
  "inputParameters.0.input": [
    "start",
    "String",
  ],
  "inputParameters.1.input": [
    "llm_test",
    "Object",
  ],
  "inputParameters.10.input": [
    "merge_0",
    "test_group_1",
  ],
  "inputParameters.11.input": [
    "merge_0",
    "test_group_2",
  ],
  "inputParameters.12.input": [
    "loop_0",
    "loop_output_var_string",
  ],
  "inputParameters.13.input": [
    "loop_0",
    "loop_output_string",
  ],
  "inputParameters.14.input": [
    "loop_0",
    "loop_output_image",
  ],
  "inputParameters.2.input": [
    "start",
    "Drilldown",
    "Object",
    "Image",
  ],
  "inputParameters.3.input": [
    "llm_test",
    "Object",
    "Number",
  ],
  "inputParameters.4.input": [
    "start",
    "Array<Code>",
  ],
  "inputParameters.5.input": [
    "llm_test",
    "Array<String>",
  ],
  "inputParameters.6.input": [
    "start",
    "Array<Object>",
    "Number",
  ],
  "inputParameters.7.input": [
    "llm_test",
    "Array<String>",
    "Number",
  ],
  "inputParameters.8.input": [
    "start",
    "Drilldown",
    "Object",
    "Undefined",
  ],
}
`;

exports[`test workflow node variables entity data > test workflow-node-input-variables-data 1`] = `
[
  [
    "start",
    [],
    [],
  ],
  [
    "end",
    [
      {
        "input": {
          "content": {
            "keyPath": [
              "start",
              "String",
            ],
          },
          "type": "ref",
        },
        "name": "start__String",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "llm_0",
              "Object",
            ],
          },
          "type": "ref",
        },
        "name": "llm_0__Object",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "start",
              "Object",
              "Image",
            ],
          },
          "type": "ref",
        },
        "name": "start__Object__Image",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "llm_0",
              "Object",
              "Number",
            ],
          },
          "type": "ref",
        },
        "name": "llm_0__Object__Number",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "start",
              "Array<Code>",
            ],
          },
          "type": "ref",
        },
        "name": "start__Array<Code>",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "llm_0",
              "Array<String>",
            ],
          },
          "type": "ref",
        },
        "name": "llm_0__Array<String>",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "start",
              "Array<Object>",
              "Number",
            ],
          },
          "type": "ref",
        },
        "name": "start__Array<Object>__Number",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "llm_0",
              "Array<String>",
              "Number",
            ],
          },
          "type": "ref",
        },
        "name": "llm_0__Array<String>__Number",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "start",
              "Object",
              "Undefined",
            ],
          },
          "type": "ref",
        },
        "name": "start__Object__Undefined",
      },
      {
        "input": {
          "content": {
            "keyPath": [],
          },
          "type": "ref",
        },
        "name": "",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "merge_0",
              "test_group_1",
            ],
          },
          "type": "ref",
        },
        "name": "merge_0__test_group_1",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "merge_0",
              "test_group_2",
            ],
          },
          "type": "ref",
        },
        "name": "merge_0__test_group_2",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "loop_0",
              "loop_output_var_string",
            ],
          },
          "type": "ref",
        },
        "name": "loop_0__loop_output_var_string",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "loop_0",
              "loop_output_string",
            ],
          },
          "type": "ref",
        },
        "name": "loop_0__loop_output_string",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "loop_0",
              "loop_output_image",
            ],
          },
          "type": "ref",
        },
        "name": "loop_0__loop_output_image",
      },
    ],
    [
      1,
      6,
      7,
      4,
      107,
      99,
      4,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      99,
      104,
    ],
  ],
  [
    "llm_0",
    [
      {
        "input": {
          "content": {
            "keyPath": [
              "start",
              "String",
            ],
          },
          "type": "ref",
        },
        "name": "start,String__llm_0,Object__start,Object,Image__llm_0,Object,Number__start,Array<Code>__llm_0,Array<String>__start,Array<Object>,Number__llm_0,Array<String>,Number__start,Object,Undefined__",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "llm_0",
              "Object",
            ],
          },
          "type": "ref",
        },
        "name": "start,String__llm_0,Object__start,Object,Image__llm_0,Object,Number__start,Array<Code>__llm_0,Array<String>__start,Array<Object>,Number__llm_0,Array<String>,Number__start,Object,Undefined__",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "start",
              "Object",
              "Image",
            ],
          },
          "type": "ref",
        },
        "name": "start,String__llm_0,Object__start,Object,Image__llm_0,Object,Number__start,Array<Code>__llm_0,Array<String>__start,Array<Object>,Number__llm_0,Array<String>,Number__start,Object,Undefined__",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "llm_0",
              "Object",
              "Number",
            ],
          },
          "type": "ref",
        },
        "name": "start,String__llm_0,Object__start,Object,Image__llm_0,Object,Number__start,Array<Code>__llm_0,Array<String>__start,Array<Object>,Number__llm_0,Array<String>,Number__start,Object,Undefined__",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "start",
              "Array<Code>",
            ],
          },
          "type": "ref",
        },
        "name": "start,String__llm_0,Object__start,Object,Image__llm_0,Object,Number__start,Array<Code>__llm_0,Array<String>__start,Array<Object>,Number__llm_0,Array<String>,Number__start,Object,Undefined__",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "llm_0",
              "Array<String>",
            ],
          },
          "type": "ref",
        },
        "name": "start,String__llm_0,Object__start,Object,Image__llm_0,Object,Number__start,Array<Code>__llm_0,Array<String>__start,Array<Object>,Number__llm_0,Array<String>,Number__start,Object,Undefined__",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "start",
              "Array<Object>",
              "Number",
            ],
          },
          "type": "ref",
        },
        "name": "start,String__llm_0,Object__start,Object,Image__llm_0,Object,Number__start,Array<Code>__llm_0,Array<String>__start,Array<Object>,Number__llm_0,Array<String>,Number__start,Object,Undefined__",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "llm_0",
              "Array<String>",
              "Number",
            ],
          },
          "type": "ref",
        },
        "name": "start,String__llm_0,Object__start,Object,Image__llm_0,Object,Number__start,Array<Code>__llm_0,Array<String>__start,Array<Object>,Number__llm_0,Array<String>,Number__start,Object,Undefined__",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "start",
              "Object",
              "Undefined",
            ],
          },
          "type": "ref",
        },
        "name": "start,String__llm_0,Object__start,Object,Image__llm_0,Object,Number__start,Array<Code>__llm_0,Array<String>__start,Array<Object>,Number__llm_0,Array<String>,Number__start,Object,Undefined__",
      },
      {
        "input": {
          "content": {
            "keyPath": [],
          },
          "type": "ref",
        },
        "name": "start,String__llm_0,Object__start,Object,Image__llm_0,Object,Number__start,Array<Code>__llm_0,Array<String>__start,Array<Object>,Number__llm_0,Array<String>,Number__start,Object,Undefined__",
      },
    ],
    [
      1,
      undefined,
      7,
      undefined,
      107,
      undefined,
      4,
      undefined,
      undefined,
      undefined,
    ],
  ],
  [
    "merge_0",
    [],
    [],
  ],
  [
    "llm_2",
    [],
    [],
  ],
  [
    "llm_3",
    [],
    [],
  ],
  [
    "156471",
    [],
    [],
  ],
  [
    "loop_0",
    [
      {
        "input": {
          "content": {
            "keyPath": [
              "start",
              "String",
            ],
          },
          "type": "ref",
        },
        "name": "loop_batch_start,String__llm_0,Object__start,Object,Image__llm_0,Object,Number__start,Array<Code>__llm_0,Array<String>__start,Array<Object>,Number__llm_0,Array<String>,Number__start,Object,Undefined__",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "llm_0",
              "Object",
            ],
          },
          "type": "ref",
        },
        "name": "loop_batch_start,String__llm_0,Object__start,Object,Image__llm_0,Object,Number__start,Array<Code>__llm_0,Array<String>__start,Array<Object>,Number__llm_0,Array<String>,Number__start,Object,Undefined__",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "start",
              "Object",
              "Image",
            ],
          },
          "type": "ref",
        },
        "name": "loop_batch_start,String__llm_0,Object__start,Object,Image__llm_0,Object,Number__start,Array<Code>__llm_0,Array<String>__start,Array<Object>,Number__llm_0,Array<String>,Number__start,Object,Undefined__",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "llm_0",
              "Object",
              "Number",
            ],
          },
          "type": "ref",
        },
        "name": "loop_batch_start,String__llm_0,Object__start,Object,Image__llm_0,Object,Number__start,Array<Code>__llm_0,Array<String>__start,Array<Object>,Number__llm_0,Array<String>,Number__start,Object,Undefined__",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "start",
              "Array<Code>",
            ],
          },
          "type": "ref",
        },
        "name": "loop_batch_start,String__llm_0,Object__start,Object,Image__llm_0,Object,Number__start,Array<Code>__llm_0,Array<String>__start,Array<Object>,Number__llm_0,Array<String>,Number__start,Object,Undefined__",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "llm_0",
              "Array<String>",
            ],
          },
          "type": "ref",
        },
        "name": "loop_batch_start,String__llm_0,Object__start,Object,Image__llm_0,Object,Number__start,Array<Code>__llm_0,Array<String>__start,Array<Object>,Number__llm_0,Array<String>,Number__start,Object,Undefined__",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "start",
              "Array<Object>",
              "Number",
            ],
          },
          "type": "ref",
        },
        "name": "loop_batch_start,String__llm_0,Object__start,Object,Image__llm_0,Object,Number__start,Array<Code>__llm_0,Array<String>__start,Array<Object>,Number__llm_0,Array<String>,Number__start,Object,Undefined__",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "llm_0",
              "Array<String>",
              "Number",
            ],
          },
          "type": "ref",
        },
        "name": "loop_batch_start,String__llm_0,Object__start,Object,Image__llm_0,Object,Number__start,Array<Code>__llm_0,Array<String>__start,Array<Object>,Number__llm_0,Array<String>,Number__start,Object,Undefined__",
      },
      {
        "input": {
          "content": {
            "keyPath": [
              "start",
              "Object",
              "Undefined",
            ],
          },
          "type": "ref",
        },
        "name": "loop_batch_start,String__llm_0,Object__start,Object,Image__llm_0,Object,Number__start,Array<Code>__llm_0,Array<String>__start,Array<Object>,Number__llm_0,Array<String>,Number__start,Object,Undefined__",
      },
      {
        "input": {
          "content": {
            "keyPath": [],
          },
          "type": "ref",
        },
        "name": "loop_batch_start,String__llm_0,Object__start,Object,Image__llm_0,Object,Number__start,Array<Code>__llm_0,Array<String>__start,Array<Object>,Number__llm_0,Array<String>,Number__start,Object,Undefined__",
      },
    ],
    [
      1,
      6,
      7,
      4,
      107,
      99,
      4,
      undefined,
      undefined,
      undefined,
    ],
  ],
  [
    "llm_in_loop_0",
    [],
    [],
  ],
  [
    "llm_in_loop_1",
    [],
    [],
  ],
]
`;

exports[`test workflow node variables entity data > test workflow-node-output-variables-data 1`] = `
[
  [
    "start",
    [
      1,
      2,
      3,
      4,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
      13,
      14,
      15,
      16,
      17,
      18,
      19,
      99,
      100,
      101,
      102,
      103,
      104,
      105,
      106,
      107,
      108,
      109,
      110,
      111,
      112,
      113,
      114,
      115,
      116,
    ],
  ],
  [
    "end",
    [],
  ],
  [
    "llm_0",
    [
      1,
      2,
      3,
      4,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
      13,
      14,
      15,
      16,
      17,
      18,
      19,
      99,
      100,
      101,
      102,
      103,
      104,
      105,
      106,
      107,
      108,
      109,
      110,
      111,
      112,
      113,
      114,
      115,
      116,
    ],
  ],
  [
    "merge_0",
    [
      undefined,
      undefined,
    ],
  ],
  [
    "llm_2",
    [],
  ],
  [
    "llm_3",
    [],
  ],
  [
    "156471",
    [],
  ],
  [
    "loop_0",
    [
      undefined,
      99,
      104,
    ],
  ],
  [
    "llm_in_loop_0",
    [
      1,
      2,
      3,
      4,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
      13,
      14,
      15,
      16,
      17,
      18,
      19,
      99,
      100,
      101,
      102,
      103,
      104,
      105,
      106,
      107,
      108,
      109,
      110,
      111,
      112,
      113,
      114,
      115,
      116,
    ],
  ],
  [
    "llm_in_loop_1",
    [],
  ],
]
`;

exports[`test workflow node variables entity data > test workflow-node-ref-variables-data 1`] = `
[
  [
    "start",
    {},
    [],
    false,
  ],
  [
    "end",
    {
      "inputParameters.0.input": [
        "start",
        "String",
      ],
      "inputParameters.1.input": [
        "llm_0",
        "Object",
      ],
      "inputParameters.10.input": [
        "merge_0",
        "test_group_1",
      ],
      "inputParameters.11.input": [
        "merge_0",
        "test_group_2",
      ],
      "inputParameters.12.input": [
        "loop_0",
        "loop_output_var_string",
      ],
      "inputParameters.13.input": [
        "loop_0",
        "loop_output_string",
      ],
      "inputParameters.14.input": [
        "loop_0",
        "loop_output_image",
      ],
      "inputParameters.2.input": [
        "start",
        "Object",
        "Image",
      ],
      "inputParameters.3.input": [
        "llm_0",
        "Object",
        "Number",
      ],
      "inputParameters.4.input": [
        "start",
        "Array<Code>",
      ],
      "inputParameters.5.input": [
        "llm_0",
        "Array<String>",
      ],
      "inputParameters.6.input": [
        "start",
        "Array<Object>",
        "Number",
      ],
      "inputParameters.7.input": [
        "llm_0",
        "Array<String>",
        "Number",
      ],
      "inputParameters.8.input": [
        "start",
        "Object",
        "Undefined",
      ],
    },
    [
      1,
      6,
      7,
      4,
      107,
      99,
      4,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      99,
      104,
    ],
    false,
  ],
  [
    "llm_0",
    {
      "inputParameters.0.input": [
        "start",
        "String",
      ],
      "inputParameters.1.input": [
        "llm_0",
        "Object",
      ],
      "inputParameters.2.input": [
        "start",
        "Object",
        "Image",
      ],
      "inputParameters.3.input": [
        "llm_0",
        "Object",
        "Number",
      ],
      "inputParameters.4.input": [
        "start",
        "Array<Code>",
      ],
      "inputParameters.5.input": [
        "llm_0",
        "Array<String>",
      ],
      "inputParameters.6.input": [
        "start",
        "Array<Object>",
        "Number",
      ],
      "inputParameters.7.input": [
        "llm_0",
        "Array<String>",
        "Number",
      ],
      "inputParameters.8.input": [
        "start",
        "Object",
        "Undefined",
      ],
    },
    [
      1,
      undefined,
      7,
      undefined,
      107,
      undefined,
      4,
      undefined,
      undefined,
    ],
    false,
  ],
  [
    "merge_0",
    {
      "groups.0.variables.1": [
        "start",
        "Object",
        "Undefined",
      ],
      "groups.0.variables.2": [
        "llm_0",
        "Array<String>",
        "Number",
      ],
      "groups.0.variables.3": [
        "start",
        "Array<Object>",
        "Number",
      ],
      "groups.0.variables.4": [
        "llm_0",
        "Array<String>",
      ],
      "groups.0.variables.5": [
        "start",
        "Array<Code>",
      ],
      "groups.0.variables.6": [
        "llm_0",
        "Object",
        "Number",
      ],
      "groups.0.variables.7": [
        "start",
        "Object",
        "Image",
      ],
      "groups.0.variables.8": [
        "llm_0",
        "Object",
      ],
      "groups.0.variables.9": [
        "start",
        "String",
      ],
      "groups.1.variables.1": [
        "start",
        "Object",
        "Undefined",
      ],
      "groups.1.variables.2": [
        "llm_0",
        "Array<String>",
        "Number",
      ],
      "groups.1.variables.3": [
        "start",
        "Array<Object>",
        "Number",
      ],
      "groups.1.variables.4": [
        "llm_0",
        "Array<String>",
      ],
      "groups.1.variables.5": [
        "start",
        "Array<Code>",
      ],
      "groups.1.variables.6": [
        "llm_0",
        "Object",
        "Number",
      ],
      "groups.1.variables.7": [
        "start",
        "Object",
        "Image",
      ],
      "groups.1.variables.8": [
        "llm_0",
        "Object",
      ],
      "groups.1.variables.9": [
        "start",
        "String",
      ],
    },
    [
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
    ],
    false,
  ],
  [
    "llm_2",
    {},
    [],
    false,
  ],
  [
    "llm_3",
    {},
    [],
    false,
  ],
  [
    "156471",
    {},
    [],
    false,
  ],
  [
    "loop_0",
    {
      "inputs.inputParameters.0.input": [
        "start",
        "String",
      ],
      "inputs.inputParameters.1.input": [
        "llm_0",
        "Object",
      ],
      "inputs.inputParameters.2.input": [
        "start",
        "Object",
        "Image",
      ],
      "inputs.inputParameters.3.input": [
        "llm_0",
        "Object",
        "Number",
      ],
      "inputs.inputParameters.4.input": [
        "start",
        "Array<Code>",
      ],
      "inputs.inputParameters.5.input": [
        "llm_0",
        "Array<String>",
      ],
      "inputs.inputParameters.6.input": [
        "start",
        "Array<Object>",
        "Number",
      ],
      "inputs.inputParameters.7.input": [
        "llm_0",
        "Array<String>",
        "Number",
      ],
      "inputs.inputParameters.8.input": [
        "start",
        "Object",
        "Undefined",
      ],
      "inputs.variableParameters.0.input": [
        "start",
        "String",
      ],
      "inputs.variableParameters.1.input": [
        "llm_0",
        "Object",
      ],
      "inputs.variableParameters.2.input": [
        "start",
        "Object",
        "Image",
      ],
      "inputs.variableParameters.3.input": [
        "llm_0",
        "Object",
        "Number",
      ],
      "inputs.variableParameters.4.input": [
        "start",
        "Array<Code>",
      ],
      "inputs.variableParameters.5.input": [
        "llm_0",
        "Array<String>",
      ],
      "inputs.variableParameters.6.input": [
        "start",
        "Array<Object>",
        "Number",
      ],
      "inputs.variableParameters.7.input": [
        "llm_0",
        "Array<String>",
        "Number",
      ],
      "inputs.variableParameters.8.input": [
        "start",
        "Object",
        "Undefined",
      ],
      "outputs.0.input": [
        "loop_0",
        "loop_var_llm_0__Object",
      ],
      "outputs.1.input": [
        "llm_in_loop_0",
        "String",
      ],
      "outputs.2.input": [
        "llm_in_loop_0",
        "Object",
        "Image",
      ],
    },
    [
      1,
      6,
      7,
      4,
      107,
      99,
      4,
      undefined,
      undefined,
      1,
      6,
      7,
      4,
      107,
      99,
      4,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
    ],
    false,
  ],
  [
    "llm_in_loop_0",
    {},
    [],
    false,
  ],
  [
    "llm_in_loop_1",
    {},
    [],
    false,
  ],
]
`;
